### ISSUE TITLE
Access-Control-Allow-Origin for HTTP-Response "302 Redirect"

### ISSUE THREAD

**Issue by kettenbach-it** [travisghansen/external-auth-server#57]
To get eas working with a Angular SPA, Javascript needs to access the HTTP-Response that eas gives to redirect the user in case of a invalid or expired session.

Would it be possible to enable eas to set Access-Control-Allow-Origin to "*".

**Comment by travisghansen** [comment]
@kettenbach-it hmm, this might be interesting. Which response(s) are you referring to?

**Comment by travisghansen** [comment]
Oh, nevermind, I see you included that in the title...I think it's possible but I'm not sure Angular will actually redirect the browser when it receives a `302` as the result of an ajax request.

This is generally a fundamental flaw with using forward auth design in conjunction with a SPA. If you have any other ideas I'd love to hear how we can possibly make the experience better.

**Comment by kettenbach-it** [comment]
No Angular will not do that by default.
But we are working on a HTTP Interceptor to exactly do that.
My colleague Tobias will join this ticket in a couple of minutes, because the HTTP Interceptor is this job.

**Comment by travisghansen** [comment]
Ah, I see! Yeah, if the application itself has knowledge then for sure it could be a much better experience.

Happy to hear feedback and I'll do some more thinking on it from my side as well considering any security implications.

You *may* be able to do this already using the custom header logic actually but I'll have to re-evaluate the code to make sure.

As an aside, if you're planning on making the app 'aware' of this kind of stuff you may consider make the app oidc aware itself and then protect the backing service with the `jwt` plugin instead of using the `oidc` plugin.

**Comment by kettenbach-it** [comment]
> As an aside, if you're planning on making the app 'aware' of this kind of stuff you may consider make the app oidc aware itself and then protect the backing service with the `jwt` plugin instead of using the `oidc` plugin.

This is exactly what we do not want to do.
The application should not care about oidc, nor should parts of the application (login-form etc) be accessible without Auth.
Therefore we let the whole application "behind" traefik and eas "disappear".
In case the eas session is over, the application has to know this and trigger a reload, which leads to a re-login at Google.

**Comment by t-itzel** [comment]
Hi Travis,

The error raised by the browser is the following:

> Access to XMLHttpRequest at 'https://dev.api.url/api/entity/1234' from origin 'http://localhost.dev.api.url/api/entity/1234:4200' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

In case of CORS, the browser blocks any response if the headers are invalid. This also means, that i am not able to detect the response which includes the redirect to the login page from Google.

I think that, when Access-Control-Allow-Origin header is set to the response including the 302 - redirect to login page, the browser will then allow the response to be parsed which means that we are able to do what we have to do.

**Comment by travisghansen** [comment]
Yeah, understood. Just wanted to make you aware of the option basically. Sounds like your team has a solid plan!

**Comment by travisghansen** [comment]
@t-itzel Yeah, your backend service actually returns this header then when the request is successfully authenticated correct?

**Comment by t-itzel** [comment]
correct, this headers is set in our backend services

**Comment by travisghansen** [comment]
Yeah ok. This lends itself quite well into my '1.0' initiatives which are to allow complete control to manipulate the response from `eas` before sending it back to the proxy. Let me put some thought into a design for this. I don't know if in the interim `traefik` will let you inject it possibly. Otherwise if you're running the code outside the containers it would be pretty easy to hack something in place.

**Comment by kettenbach-it** [comment]
AFAIK I can allow traefik to do that: https://docs.traefik.io/middlewares/forwardauth/#authresponseheaders

We are using traefik 1.7.

EAS 1.0 is becoming a very interesting alternative to https://cloud.google.com/iap. Maybe you can do al little SEO in the README - if you want more people to use it.

**Comment by travisghansen** [comment]
Those headers in `traefik` is actually what `traefik` will pass to the backing service in the case of success. So won't work for this scenario.

I will actually hack the feature together as a sister feature to https://github.com/travisghansen/external-auth-server/blob/master/HEADERS.md

I'll commit shortly, but I made the docker builds multi-arch now so it takes a while for them to build :(

I've generally not been pro-actively pushing it due to known breaking changes for '1.0'. I'm not trying to hide the project either, but once I hit '1.0' I'll be a little more pro-active about 'marketing' it.

**Comment by travisghansen** [comment]
K, here ya go: https://github.com/travisghansen/external-auth-server/runs/475412255?check_suite_focus=true

When that finishes building just do the following:

```
          "custom_error_headers": {
            "Access-Control-Allow-Origin": {
              "source": "static",
              "query_engine": "static",
              "query": "*"
            }
          },
```

**Comment by kettenbach-it** [comment]
@t-itzel I've build the container and startet the service with custom header "Access-Control-Allow-Origin". Please try who the SPA now behaves.

**Comment by kettenbach-it** [comment]
@travisghansen Would it be possible to set the custom header depending on the original request?

The reason is, that if I put "*" in the header, the browser won't send the credentials and our SPA won't get access to the reply.
And if I set the header to a fixed value, the browser will send credentials but I have to have separate config_tokens for each (sub-)domain - each containing different Access-Control-Allow-Origin custom-headers. If I'm not mistaken, I'd loose the single sign on that I'd like to implement, which would be too bad.

The ideal way to send back the Access-Control-Allow-Origin with the value set to the origin domain, that the SPA is running on.

Something like:

```
 "custom_error_headers": {
            "Access-Control-Allow-Origin": {
              "source": "static",
              "query_engine": "static",
              "query": $originDomain
            }
          }
```

would be a very cool  feature.

**Comment by travisghansen** [comment]
Yes actually, there are other places I'm doing this kind of thing. Let me dig through the code again and see if it's currently possible in this context.

**Comment by kettenbach-it** [comment]
I think this might be already possible using the X-Forwarded-Host set by traefik in each request.

**Comment by travisghansen** [comment]
Yeah, I have a custom internal data struct that gets passed around which includes the 'parentRequestInfo' normalized to work with all the different proxies.

This same idea is used for dynamic server-side config tokens.

**Comment by kettenbach-it** [comment]
After a few days of working with eas I realize how infinitely flexible this software is. At first I was irritated by the unusual way of configuring it - I have never seen the concept with the encrypted config_tokens before. Whereby I have to say that I am running a server-side config passed in environment variables, as I think that because of the seperation of concerns the config belongs in the esa-container and not in the backend service. But as I said: the system is very flexible and everybody can use it as he wants.
If it works now to solve the annoying CORS issue - that would be great.

**Comment by travisghansen** [comment]
Yeah, the approach is definitely very unique. By the time I implement the '1.0' stuff there will be much more power available (meaning, I'll make it possible to completely alter responses in a very generic/powerful fashion). Right now the specific operating datasets for assertions/headers/etc are too limited and don't allow cross-dataset assertions (ie: assert something in the token against some header in the original request, or assert something in the `idToken` against `userinfo` etc). My intent is to make 1 data struct that includes everything, and then give complete control over manipulating the response. Each plugin will be responsible for returning back to the main controller `authData` specific to the plugin and it simply becomes part of the post processing dataset.

You can find more details here in #29 

In any case, I'll dig through the code again and get back to you shortly about whether the current header inject stuff has that kind of access.

**Comment by travisghansen** [comment]
OK, available here: https://github.com/travisghansen/external-auth-server/commit/d352f221ecdf05b93f436bec6303e477b5b38175

```
          "custom_error_headers": {
            "Access-Control-Allow-Origin": {
              "source": "parentRequestInfo",
              "query_opts": {
                "single_value": true
              },
              "query_engine": "jp",
              "query": "$.parsedUri.host"
            }
          },
```

Let me know how it works out.

**Comment by travisghansen** [comment]
Actually, that's probably still not what you're after...that returns the domain of the API, not the SPA :(

**Comment by travisghansen** [comment]
Yeah, if you want the domain of the SPA do this:

```
          "custom_error_headers": {
            "Access-Control-Allow-Origin": {
              "source": "req",
              "query_opts": {
                "single_value": true
              },
              "query_engine": "jp",
              "query": "$.headers.origin"
            }
          },
```

Note that it will be empty for non-XHR requests though I'd assume.

**Comment by kettenbach-it** [comment]
The of supplying customer error headers with CORS fields now works. The SPA is able to parse the reply as intended!

Now the problem is that the browser immediately executes the 302 and calls accounts.google.com Since their server does not set CORS headers, we have the same problem as before: the SPA is denied access because of no Access-Control-Allow-Origin being present. Unfortunately you can't prevent the redirect, let alone intercept it in any way.

idea:
Is there any chance that eas will only return the 302 if the call came from the browser itself instead of the Angular service? We could set a header in Angular that identifies the SPA. Or we could filter by the origin-header as we did with the CORS headers - that would be fine too.

If the conditions are met, then another http status code (e.g. 399) with defined payload is sent instead of the 302. For example, you could use the location field used for the 3 errors to send the browser the redirect URL. Provided the browser does not redirect at 399 and Location. Then you could call the field payload or something.
I can then intercept and react to this in Angular.
So:
If the user reloads or calls the app in the browser, the header is missing in the request and Traefik normally sends http status 302 with redirect url.
If the user is in the app and the Angular Service makes an XHR on the api, then the header is set and Traefik sends a http status 399 with payload { status: 302, redirect_url: string }

Wow, this is pretty advanced. I wonder how others do this SPA. If that works, eas is the perfect service for SPA. 
I also agree to document this and to provide my config as an example.

**Comment by travisghansen** [comment]
Yeah, you're really pointing out why I just need to sit down and hammer out my '1.0' ideas.

In any case, in the interim if you're willing to use the `request_js` plugin (https://github.com/travisghansen/external-auth-server/blob/master/PLUGINS.md#request_js) then I think the easiest way to do this would be to declare that as the 2nd plugin in the pipeline. In order to facilitate I would need to extend slightly the internal API to ensure the `request_js` plugin gets access to the `res`ponse from the previous plugin in the pipeline (assuming it follows another plugin).

In pseudo code for the `request_js` plugin:

```
if (previousPluginResponse.statusCode == 30X && req.headers.orgin) {
  res.statusCode = 399;
  res.body('some data');
  res.Header('foo', 'bar');
  return res;
} else {
  return previousPluginResponse;
}
```

Review the doc on `request_js` and then let me know if conceptually that works for you. You may want to confirm the browser behaves the way you want it to when receiving whatever status code you want to use. If it does work, it shouldn't be too bad to refactor the internals to send the `previousPluginResponse` to the plugins...specifically `request_js`. It will generally be very simple to extend.

SPAs generally have direct knowledge of `oidc` rather than using the forward auth pattern. I find the route you're taking very interesting actually and think it is a pretty cool idea. I'm interested to see how it all works in the end.

**Comment by travisghansen** [comment]
Another potential option would be to add an option to `oidc` plugin that allows to define the status code for redirects when it's an ajax request. For example send a `403` but you'd still have the `Location` header. I actually quite like this idea... thoughts?

**Comment by kettenbach-it** [comment]
> SPAs generally have direct knowledge of `oidc` rather than using the forward auth pattern. I find the route you're taking very interesting actually and think it is a pretty cool idea. I'm interested to see how it all works in the end.

We just try to do it like Google. Their GSuite apps are hidden behind OAuth like our app.

**Comment by kettenbach-it** [comment]
> Another potential option would be to add an option to `oidc` plugin that allows to define the status code for redirects when it's an ajax request. For example send a `403` but you'd still have the `Location` header. I actually quite like this idea... thoughts?

Tobias suggested this as well the other day. Would definitiv work for us.

**Comment by kettenbach-it** [comment]
> In any case, in the interim if you're willing to use the `request_js` plugin (https://github.com/travisghansen/external-auth-server/blob/master/PLUGINS.md#request_js) then I think the easiest way to do this would be to declare that as the 2nd plugin in the pipeline. In order to facilitate I would need to extend slightly the internal API to ensure the `request_js` plugin gets access to the `res`ponse from the previous plugin in the pipeline (assuming it follows another plugin).
> 

Would be an option, too, and we could do this.
But I get the impression that you've already given yourself the answer. Forcing a different status code in the `oidc` plugin for SPA related redirects would be a very simple and elegant solution. It would another great feature to `eas`and it would be a lot simpler then the generic `request_js`

**Comment by travisghansen** [comment]
Yeah, I'll add some options to control that directly in the plugin.

**Comment by travisghansen** [comment]
Ok, added in commt https://github.com/travisghansen/external-auth-server/commit/a119f8b63157cb280f45b1de2145724d8b944db4 and documented in https://github.com/travisghansen/external-auth-server/blob/master/PLUGINS.md

Look for the `xhr` key.

I've also finally tagged `v0.7.0` which includes this feature so it's now in a tagged release.

**Comment by t-itzel** [comment]
Works great :) 👍 

One thing is left: 

The url of the original xhr-request is encoded as redirect_uri within the response location (accounts.google.com...). In our case this must rather be the referrer of the original xhr-request.

Example:
Browser-Url is https://app.domain.com/job/123 which inititiates 2 api xhr-requests:
- https://app.domain.com/api/info
- https://app.domain.com/api/job/123
Both requests end in a 401-response but with different redirect_urls. 

**Current behavior**
If the first response location is used, the user is redirected to https://app.domain.com/api/info after successful login.
If the first response location is used, the user is redirected to https://app.domain.com/api/job/123 after successful login.

**Desired behavior**
The user will be redirected to the original browser url https://app.domain.com/job/123 after a successful login.

**Comment by travisghansen** [comment]
Oh interesting. Does `origin` contain path info? Where do you propose we obtain that data from?

**Comment by t-itzel** [comment]
Ah yes, I forgot. It's the `referer`, you're right.

**Comment by travisghansen** [comment]
Implemented here: https://github.com/travisghansen/external-auth-server/commit/096db7d89d017698f2e925f70c8c0d899f9f6d5d

In the `next` branch. Images will take a while to build.

**Comment by kettenbach-it** [comment]
First tests show, that it works as expected with all kinds of URLs, including the longest we have (which are not very long: 2 chars as path).

Let's keep that issue open for 1-2 days. We will gain experience with it.

**Comment by kettenbach-it** [comment]
With an url path of: /jobs/details/10757/123/456/789/0/abc/def/ghi/jkl
where starting with 123 the url artificially made, I get an error:

![image](https://user-images.githubusercontent.com/7711872/75678114-276fbd80-5c8d-11ea-9e2c-ccb5fef0b665.png)

Where is the limit for these redirect url?

**Comment by travisghansen** [comment]
Hmm, that's interesting. Couple questions:

1. Have you explicitly set the `redirect_uri` option in your token?
1. Have you tried the same request in an incognito browser?

**Comment by kettenbach-it** [comment]
1. The redirect_uri is set in my config_token:
```
/**
                 * static redirect URI
                 * if your oauth provider does not support wildcards place the URL configured in the provider (that will return to this proper service) here
                 */
                redirect_uri: "https://auth.mydomain.com/oauth/callback",
```
2. I tried: it happens with incognito browser as well

**Comment by travisghansen** [comment]
Ok. I use the `state` parameter to do some magical stuff with `eas`. If you remove that param from the authentication URL does it load the login screen? If so the only way I can think of to mitigate the issue would be to store the state server side and make the state simply be an ID that gets passed around.

**Comment by travisghansen** [comment]
Also, just so I'm clear, you only get this error when going to a very long url initially right? If you shorten the url all works as expected correct?

**Comment by kettenbach-it** [comment]
1. Yes, I only get this error with a long (artificial) url! With "normal" urls everything works (for now).

2. a) Without the static `redirect_uri`, the redirect is done to the originating url "inside" the SPA. And since Google does not support wildcards in in the redirect URIs, this is not an option. Otherwise I had to put every single route of our SPA in the list of authorized redirect urls, which is not practicable. We need to redirect to eas. 
b) You could listen to something like http://app.mydomain.com/oauth/callback and put this as a redirect url. traefik-forward-auth of thomseddon offers this option. But actually the same as 2a and the solution with one static url is even more comfortable.

3. I think it makes sense to store the state on the server and just pass an id, because we do not rely on whether and how different services do the redirect. I have stumbled over too long urls etc. at researches to the topic already the one or other time

**Comment by travisghansen** [comment]
Very good. Yes I much prefer how this project handles the redirect uris generally to alternatives. The direct endpoint that handles that is what actually uses the state data as I embed the 'real' redirect in the state. This allows the single endpoint to service infinite # of apps including even cross domain.

If you remove the `state` query parameter from the auth url does the login screen load?

**Comment by kettenbach-it** [comment]
It is a bit tricky to change the very long url manually. Also, after I send the url the first time, there is a redirect to the account login again. If I remove the state again parameter, eas.

`2020-03-02T16:29:32.914854425Z error: The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type undefined {"stack":"Error: The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type undefined\n    at Object.decrypt (/home/eas/app/src/utils.js:48:11)\n    at server.WebServer.get (/home/eas/app/src/plugin/oauth/index.js:232:36)\n    at Layer.handle [as handle_request] (/home/eas/app/node_modules/express/lib/router/layer.js:95:5)\n    at next (/home/eas/app/node_modules/express/lib/router/route.js:137:13)\n    at Route.dispatch (/home/eas/app/node_modules/express/lib/router/route.js:112:3)\n    at Layer.handle [as handle_request] (/home/eas/app/node_modules/express/lib/router/layer.js:95:5)\n    at /home/eas/app/node_modules/express/lib/router/index.js:281:22\n    at Function.process_params (/home/eas/app/node_modules/express/lib/router/index.js:335:12)\n    at next (/home/eas/app/node_modules/express/lib/router/index.js:275:10)\n    at middleware (/home/eas/app/node_modules/express-prom-bundle/src/index.js:164:5)\n    at Layer.handle [as handle_request] (/home/eas/app/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/home/eas/app/node_modules/express/lib/router/index.js:317:13)\n    at /home/eas/app/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/home/eas/app/node_modules/express/lib/router/index.js:335:12)\n    at next (/home/eas/app/node_modules/express/lib/router/index.js:275:10)\n    at cookieParser (/home/eas/app/node_modules/cookie-parser/index.js:71:5)\n    at Layer.handle [as handle_request] (/home/eas/app/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/home/eas/app/node_modules/express/lib/router/index.js:317:13)\n    at /home/eas/app/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/home/eas/app/node_modules/express/lib/router/index.js:335:12)\n    at next (/home/eas/app/node_modules/express/lib/router/index.js:275:10)\n    at urlencodedParser (/home/eas/app/node_modules/body-parser/lib/types/urlencoded.js:91:7)"}`

I'm not 100% sure if this what you wanted.

**Comment by travisghansen** [comment]
I guess I would have expected that. Mostly wanted to know if Google auth page actually loads and let's you attempt to sign in..from what I can tell yes.

**Comment by travisghansen** [comment]
Could you send over the full auth URL? I've put in a very large url for a protected path and I'm still not hitting the/a limit with google..

**Comment by kettenbach-it** [comment]
URLs send by mail

**Comment by travisghansen** [comment]
Wow, that is definitely a massive URL and over the generally accpted limit of 8096 or so for chars.

I'm a bit perplexed by that, how many characters are your issuer sign/encrypt secrets? I wonder if the length of those somehow impacts the data length..

**Comment by kettenbach-it** [comment]
The secrets are all 20 chars.

**Comment by travisghansen** [comment]
Did you truncate the query string on the originally requested URL?

**Comment by kettenbach-it** [comment]
You mean the request of app.mydomain.com? No!

**Comment by travisghansen** [comment]
Yeah, something is really odd here for sure. For example, I have a URL that is 849 characters long that produces a login URL that 3064. Your login URL is 11580 chars with an original URL of 89 chars.

The data being encrypted/signed is static in length with the exception of the URI to return the user back to after successful login.

https://github.com/travisghansen/external-auth-server/blob/master/src/plugin/oauth/index.js#L342

Are you running `eas` locally where you can hack in some extra debug lines?

**Comment by kettenbach-it** [comment]
Yes!

**Comment by travisghansen** [comment]
Can you log out the `payload` `stateToken` and `state` right before the `const url = await plugin.get_authorization_url` line near the link above?

**Comment by travisghansen** [comment]
Maybe you have many many `scopes`? Or a bunch of `custom_authorization_parameters`?

**Comment by kettenbach-it** [comment]
```
scopes: ["openid", "email", "profile"], // must include openid
                // custom static authorization URL parameters
                // NOTE: all critical fields are managed automatically, this should only be used in advanced scenarios
                // ie: https://developers.google.com/identity/protocols/OpenIDConnect#refresh-tokens
                // https://github.com/travisghansen/external-auth-server/issues/51
                custom_authorization_parameters: {
                    'access_type': 'offline',
                    "prompt": "consent"
                },
```

**Comment by kettenbach-it** [comment]
![image](https://user-images.githubusercontent.com/7711872/75756183-7d497180-5d30-11ea-9f51-8fbc478a8b63.png)

**Comment by travisghansen** [comment]
Yeah, I just realized those were all in the link you already sent...all the length is for sure in the state...just baffled why it's such a large payload.

And yes, you're logging in the correct spot.

**Comment by travisghansen** [comment]
Send over the values of those logs here or email them..either way.

**Comment by kettenbach-it** [comment]
Sent detailed mail

**Comment by travisghansen** [comment]
If I'm reading this right you're getting into a login loop? Meaning, you keep getting redirected back to google to authenticated and then eventually after re-authenticating X number of times you're getting the 413?

**Comment by kettenbach-it** [comment]
It loops once and then, after the 2nd login, it results in 413

**Comment by travisghansen** [comment]
What value do you have for the `sameSite` on the session cookie? Can you try disabling the `csrf` feature and see what happens?

```
          "csrf_cookie": {
            "enabled": false
          },
```

**Comment by kettenbach-it** [comment]
Both are set to Lax:

![image](https://user-images.githubusercontent.com/7711872/75759147-635e5d80-5d35-11ea-91a4-dc363e9d1676.png)


I'll try disabled csrf

**Comment by kettenbach-it** [comment]
The behaviour is the same

**Comment by travisghansen** [comment]
Disable `secure` for both of them, I'm guessing you're never getting a session since it's running over `http` instead of `https`. The browser is likely rejecting both cookies.

**Comment by kettenbach-it** [comment]
I sent the log my mail.
Why does it run over http?

**Comment by kettenbach-it** [comment]
Ah, okay. Because it's my localhost

**Comment by travisghansen** [comment]
Your development server is running over http (looks like you're running ember).

**Comment by travisghansen** [comment]
Is the second log with `secure` set to `false`?

**Comment by kettenbach-it** [comment]
With disabled secure, it is the same.
Maybe my local development server is not the right place.
But with short urls it works there as well

**Comment by kettenbach-it** [comment]
I sent a third log.
I will try it on the public server

**Comment by kettenbach-it** [comment]
In the development in localhost, we used to get a cookie by calling a public api, that the browser then uses

**Comment by travisghansen** [comment]
I'm guessing there's something else at play and the url length is just noise...can't be sure yet though. I'll probably need you to crank the log level up a bit so I can get a clearer picture of what's going on.

From the info I can gather from the logs it's reaching a state where it's trying to authenticate the special endpoint which handles authentication...it's *really* strange.

**Comment by kettenbach-it** [comment]
On the public url it works. See mail with detailed log.

We could definitly live with it, if on our laptops, we can't directly login to the app at localhost, but need to get a cookie from the (public) api (we have a dev-api publicly for all developers).

**Comment by kettenbach-it** [comment]
In public it works with a even much longer url:

https://app.mydomain.com/jobs/details/10757/123/456/789/0/abc/def/ghi/jkl/123/456/789/10757/123/456/789/0/abc/def/ghi/jkl/123/456/789

**Comment by travisghansen** [comment]
You should be able to run it locally as well. Do you run the API locally from your development machines as well? I think if you send over `debug`/`silly` level logging I'll be able to wrap my brain around what's going on.

**Comment by kettenbach-it** [comment]
The API for the development laptops run public with https, just like the prod api but with a different database.

I sent a 5th log with silly loglevel

**Comment by travisghansen** [comment]
Ah ha! I see what's going on here. It's connected to this comment here from the plugin config:

```
//setting this generally assumes the API and UI are secured by the same eas config
```

So you definitely *are* getting into a loop with this and it's related to the the `referer` setting. I'm guessing in prod both the UI and the API are served under the same domain perhaps? or at least both are secured by `eas` using the same config?

It's **really** late for me now but I have a pretty good idea what's going on and why you're seeing what you're seeing. I *think* it's possible to solve systemically and removing that note...but we'll see. I'll revisit tomorrow when I'm less delirious :)

**Comment by kettenbach-it** [comment]
Hi Travis! Thanks a lot and good night!

While you're sleeping, I'll answer your last comment

**Comment by kettenbach-it** [comment]
Our setup is as follows:

EAS: right now all public endpoints below are secured by a single eas with a single serverside config_token.
The Cookie-Domain is set to *.mydomain.com, so it covers everything described below.

Production
- API URL: https://app.mydomain.com/api
- SPA URL: https://app.mydomain.com
two different containers being served in hierarchy by traefik, secured by eas as stated above 

Development:
- API URL: https://dev.mydomain.com/api
- SPA URL: http://localhost.dev.mydomain.com (No SSL!)
Environment on our laptops just served by "ng serv". 
localhost.dev.mydomain.com points  to 127.0.0.1 in DNS
That's the trick, how we can use the credentials on localhost and get access to the public DEV-API URL

In the heat of the moment I totally forgot, that our development system is quite different and a little bit tricky. With the solution before, without eas and the http-interceptor, the developers always had to get a valid cookie by calling the api url, signing in and then pointing the browser to http://localhost.dev....

It doesn't matter if it doesn't work in the development setup.
I already informed the team, that it's better to get a cookie the way we did it all the time, in case Google shows the broken robot.

**Comment by travisghansen** [comment]
Yeah the additional context is extremely helpful. I can make the logic better than it is currently which logically will eliminate the need for the note previously mentioned.  I can't guarantee it will work but I think it will facilitate your scenario cleanly and generally just be better logic.

As it stands currently I'm redirecting to the SPA (`referer`) after successful auth from the provider just before auth with `eas`. I need to update that logic to redirect to the `referer` after the auth with `eas` (of course based on settings etc). This is logically more sound and I'm guessing will cover your use case 100% but we'll see :) incidentally I'll need to store more data in `state` so I may also migrate that to be stored server side regardless.

Tied up in several meetings today so not sure when I'll get to it.

**Comment by travisghansen** [comment]
Try this: https://github.com/travisghansen/external-auth-server/commit/efedd8be38fbbf873fcbbe44bb7ebef6c8920e2a

I didn't make `state` server-side yet, but it should be better logic. One thing I don't know is if the browsers actually accept/store cookies sent via xhr requests...if so you should be fine, if not you'll need to disable the csrf cookie.

**Comment by kettenbach-it** [comment]
I tested the current next branch: I tried both environments "dev" and "prod" with short and long urls.
All scenarios worked fine.

I set my cookies back to secure and enabled csrf cookie:

![image](https://user-images.githubusercontent.com/7711872/75854879-63bc2e80-5df1-11ea-82f8-9e62a64a3d88.png)

What do you think: are these cookie settings correct for production?

**Comment by travisghansen** [comment]
Great! I may rename the `xhr` keys yet so beware that. I've read up quite a bit and not sure I'm totally following the standards at this point so need a bit more time to digest that..

The config seems valid. If the API is served up over secure then I think you'll be ok with that setting.

**Comment by travisghansen** [comment]
Can you run through some tests of this functionality as it stands with safari and firefox? Would be helpful to know if they behave appropriately in the various scenarios as well.

I'm *thinking* at this point that I'll add some additional logic to *try* to properly detect `xhr` requests but unfortunately it's not 100%. Generally I think it's in pretty decent shape and can certainly serve well folks who have control over the app.

For example right now based on the config you sent above I'm guessing in your `prod` setup the app is not sending the `origin` header (ie: it's not considered a `cors` request so I'm guessing the browser omits sending that). In that case, if for some reason a token expires or whatever (after the initial load of the app) then you would get redirected back to the api endpoint instead of the referrer location.

The logic path for detecting `xhr` is shaky. If `origin` is present, it's almost certainly `xhr`, however, as mentioned above if `origin` is NOT present it does not necessarily mean it is not `xhr`. I think I'll expand the logic for detecting `xhr` to also include the semi-standard `X-Requested-With: XMLHttpRequest` header which will allow folks who have control of the app to send an indicator.

**Comment by kettenbach-it** [comment]
> Can you run through some tests of this functionality as it stands with safari and firefox? Would be helpful to know if they behave appropriately in the various scenarios as well.

With firefox everything works. We regularly test with Chrome and Firefox.
With Safari Version 12.1.2 (12607.3.10) (i have a pretty old Mac OS 10.12) there are problems I sent you a detailed logfile by mail.

> The logic path for detecting `xhr` is shaky. If `origin` is present, it's almost certainly `xhr`, however, as mentioned above if `origin` is NOT present it does not necessarily mean it is not `xhr`. I think I'll expand the logic for detecting `xhr` to also include the semi-standard `X-Requested-With: XMLHttpRequest` header which will allow folks who have control of the app to send an indicator.

You could introduce a custom header that needs to sent by the SPA with every request. This way you can reliably distinguish between requests from the SPA and others.

**Comment by travisghansen** [comment]
Yep, I'll use the header mentioned previously. That serves folks who have control over the app well, but my concern is a false sense of coverage for apps being secured by `eas` which have no knowledge of `eas` (something like grafana, etc).

**Comment by kettenbach-it** [comment]
Oh, you're right. 
I am afraid that in cases without access to the SPA it will be difficult to build such a scenario.
Whereby Grafana implements afaik its own OIDC client. This will be the case for many applications of this kind, because they have to be able to run in systems without a load balancer. For example on a simple web server.
Our scenario is quite special, but could be even more widespread in the enterprise environment.

**Comment by travisghansen** [comment]
Yeah, grafana does support oidc it's just an example I know a lot of people secure with `eas` for single signon etc. In any case the pitfalls of forward auth with SPA are generally known and they won't be *worse* off with the `xht` settings than without them.

In practice the SPA issue is not very disruptive anyhow especially if tokens are being automatically refreshed with a reasonable session window for the use-case. Just want to make sure it gets documented properly so users know what they're getting into mostly.

Appreciate all the feedback! I think this is a valuable addition to the project.

**Comment by kettenbach-it** [comment]
II tried IE 11.0: the Auth seems to work okay, but the SPA does not load. So we need to check first what's wrong with loading Angular

**Comment by kettenbach-it** [comment]
Chrome 78 and Firefox 70.0.1 on Windows 7 both work fine.

**Comment by kettenbach-it** [comment]
Edge 80.0.361.66 on Windows 7 works fine (is that already a Chrome?)

**Comment by kettenbach-it** [comment]
With Internet Exploder 11.0 the login to the API  - just with the browser - works.
So that's not a matter of eas.

**Comment by travisghansen** [comment]
So if I'm following so far the only issue you've hit is when using an old safari?

**Comment by kettenbach-it** [comment]
Yes

**Comment by travisghansen** [comment]
Ok, I'll probably add the header support today and then it'll be ready to snap a release.

**Comment by travisghansen** [comment]
Just updated to `xhr` detection logic in the `next` branch to support the `X-Requested-With: XMLHttpRequest` header.

**Comment by travisghansen** [comment]
Released to `master`/`latest` and `v0.8.0`.
